### PR TITLE
feat: add is_abs_ppt function to check if a density matrix is absolut…

### LINKED ISSUE #208: Feature: Function to determine if density matrix is absolutely PPT
Similar to the `IsAbsPPT` function from QETLAB ([link](https://github.com/nathanieljohnston/QETLAB/blob/master/IsAbsPPT.m)).

This would also require implementing the `AbsPPTConstraints` or equivalent in toqito as well.

### PR DESCRIPTION
**Pull request by luffy-orf**
### Overview

This PR implements the `is_abs_ppt` function in `toqito.state_props`, providing a way to determine if a bipartite density matrix is **absolutely PPT** (i.e., PPT from spectrum), as requested in #208 and inspired by QETLAB's [IsAbsPPT](https://qetlab.com/IsAbsPPT) and [AbsPPTConstraints](https://qetlab.com/AbsPPTConstraints).

### Key Changes

- **New function:** `is_abs_ppt` for checking if a density matrix is absolutely PPT.
    - Supports 2x2, 2x3, and 3x3 bipartite systems (explicit constraint matrices).
    - Returns `True` if absolutely PPT, `False` if not, and `-1` for unsupported higher dimensions.
    - Validates input for squareness, Hermiticity, trace, and positivity.
- **Helper:** Internal function for generating constraint matrices as per Hildebrand (2007).
- **Tests:** Added tests for maximally mixed states, random states, known non-absolutely PPT states, invalid input, and large dimensions.
- **API:** Exposed `is_abs_ppt` in the public API.

Closes #208

### CODE REVIEW

File: toqito/state_props/is_abs_ppt.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    """Determine whether a density matrix is absolutely PPT (PPT from spectrum).
+
+    Args:
+        mat: The density matrix to check.
+        dim: The local dimensions (default: inferred as equal dims).
+        tol: Numerical tolerance for positive semidefinite checks.
+
+    Returns:
+        True if absolutely PPT, False if not, -1 if undecidable (for large dims).
+
```
Comment (purva-thakre) [inline toqito/state_props/is_abs_ppt.py:82]: @luffy-orf We have a different style for the docstrings. Could you take a look at the contributing guide in the documentation?

File: toqito/state_props/tests/test_is_abs_ppt.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    """Test that maximally mixed states are absolutely PPT for supported dimensions."""
+    # 2x2
+    rho_2x2 = max_mixed(4)
+    assert is_abs_ppt(rho_2x2, [2, 2]) is True
+    # 2x3
+    rho_2x3 = max_mixed(6)
+    assert is_abs_ppt(rho_2x3, [2, 3]) is True
+    # 3x3
+    rho_3x3 = max_mixed(9)
+    assert is_abs_ppt(rho_3x3, [3, 3]) is True
```
Comment (purva-thakre) [inline toqito/state_props/tests/test_is_abs_ppt.py:21]: if a unit test is following the same pattern, it is better to use https://docs.pytest.org/en/7.1.x/how-to/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions

File: toqito/state_props/tests/test_is_abs_ppt.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        is_abs_ppt(np.ones((3, 4)))
+    # Not Hermitian
+    with pytest.raises(ValueError):
+        is_abs_ppt(np.array([[0, 1], [0, 0]]))
+    # Not trace 1
+    with pytest.raises(ValueError):
+        is_abs_ppt(np.eye(4) * 2)
+    # Not positive semidefinite
+    with pytest.raises(ValueError):
+        is_abs_ppt(np.array([[1, 0], [0, -1]]))
```
Comment (purva-thakre) [inline toqito/state_props/tests/test_is_abs_ppt.py:64]: You also want to check the error message. 

Same as my other comment, use https://docs.pytest.org/en/7.1.x/how-to/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions wherever possible.

File: toqito/state_props/is_abs_ppt.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    :raises ValueError: If the input matrix is not a valid density matrix.
+    """
+    if not is_square(mat):
+        raise ValueError("Input matrix must be square.")
+    if not np.allclose(mat, mat.conj().T, atol=tol):
+        raise ValueError("Input matrix must be Hermitian.")
+    if not np.isclose(np.trace(mat), 1, atol=tol):
+        raise ValueError("Input matrix must have trace 1 (density matrix).")
+    if np.any(np.linalg.eigvalsh(mat) < -tol):
+        raise ValueError("Input matrix must be positive semidefinite.")
```
Comment (purva-thakre) [inline toqito/state_props/is_abs_ppt.py:113]: we have a function for this.

File: toqito/state_props/tests/test_is_abs_ppt.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+@pytest.mark.parametrize("dims", [([2, 2]), ([2, 3]), ([3, 3])])
+def test_random_psd_not_absolutely_ppt(dims):
+    """Test that random PSD states are not always absolutely PPT."""
+    np.random.seed(42)
+    n = dims[0] * dims[1]
+    for _ in range(3):
+        rho = random_psd_operator(n)
+        rho = rho / np.trace(rho)
+        res = is_abs_ppt(rho, dims)
+        assert res in [True, False, -1]
```
Comment (purva-thakre) [inline toqito/state_props/tests/test_is_abs_ppt.py:30]: Not the way to go about verifying this. The result is always going to be in this list. Add an additional var for the `expected_result` and then verify it.

File: toqito/state_props/is_abs_ppt.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    :param dim: The local dimensions (default: inferred as equal dims).
+    :param tol: Numerical tolerance for positive semidefinite checks.
+    :return: :code:`True` if absolutely PPT, :code:`False` if not, :code:`-1` if undecidable (for large dims).
+    :raises ValueError: If the input matrix is not square, not Hermitian, does not have trace 1, is not positive
+        semidefinite, or if the provided dimensions do not match the matrix size.
+
+    Note:
+        Returns -1 if the absolutely PPT property is undecidable for the given dimensions (i.e., for dimensions
+        other than 2x2, 2x3, 3x2, or 3x3).
+
```
Comment (purva-thakre) [inline toqito/state_props/is_abs_ppt.py:103]: You need the references code block here such that the references are listed in the docs build.

File: toqito/state_props/is_abs_ppt.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+# @article{Hildebrand_2007_PPT,
+#   author = {Roland Hildebrand},
+#   title = {Positive partial transpose from spectra},
+#   journal = {Physical Review. A},
+#   volume = {76},
+#   number = {5},
+#   year = {2007},
+#   doi = {10.1103/PHYSREVA.76.052325},
+#   url = {https://doi.org/10.1103/PHYSREVA.76.052325}
+# }
```
Comment (purva-thakre) [inline toqito/state_props/is_abs_ppt.py:155]:
```suggestion

```

File: toqito/state_props/is_abs_ppt.py [hunk 8]
```diff
@@ -78,7 +74,10 @@ def _abs_ppt_constraints(eigvals: np.ndarray, dim: list[int]) -> list[np.ndarray
 def is_abs_ppt(mat: np.ndarray, dim: None | int | list[int] = None, tol: float = 1e-8) -> bool | int:
```
Comment (vprusso) [inline toqito/state_props/is_abs_ppt.py:74]: If we are subsequently passing `tol` as `atol`, we likely should name this accordingly.

Comment (vprusso) [inline toqito/state_props/is_abs_ppt.py:74]: Should return a `bool` (but not an `int`, refer to the comment below).

File: toqito/state_props/is_abs_ppt.py [hunk 9]
```diff
@@ -78,7 +74,10 @@ def _abs_ppt_constraints(eigvals: np.ndarray, dim: list[int]) -> list[np.ndarray
...
     r"""Determine whether a density matrix is absolutely PPT (PPT from spectrum).
 
-    A density matrix is absolutely PPT if every unitary conjugation of it is PPT. This is a spectrum-based property and can be checked using the Hildebrand/QETLAB constraints for small dimensions.
+    :cite:`Hildebrand_2007_PPT, QETLAB_link`.
+
+    A density matrix is absolutely PPT if every unitary conjugation of it is PPT. This is a spectrum-based
+    property and can be checked using the Hildebrand/QETLAB constraints for small dimensions.
```
Comment (vprusso) [inline toqito/state_props/is_abs_ppt.py:80]: The constraints are from Hildebrand's paper, but the reference implementation is from QETLAB. These should be referred to as separate items.

File: toqito/state_props/is_abs_ppt.py [hunk 10]
```diff
@@ -141,3 +140,16 @@ def is_abs_ppt(mat: np.ndarray, dim: None | int | list[int] = None, tol: float =
         if not is_positive_semidefinite(L, atol=tol):
             return False
     return True
+
+
+# TODO: Add the following BibTeX entry to docs/refs.bib if not present:
```
Comment (vprusso) [inline toqito/state_props/is_abs_ppt.py]: Seems like you left a `TODO` stub here.

File: toqito/state_props/tests/test_is_abs_ppt.py [hunk 11]
```diff
@@ -1,47 +1,52 @@
 """Unit tests for the is_abs_ppt function in toqito.state_props."""
 
+import re
```
Comment (vprusso) [inline toqito/state_props/tests/test_is_abs_ppt.py:3]: Is there a reason `re` is imported? I don't see it being used.

File: toqito/state_props/tests/test_is_abs_ppt.py [hunk 12]
```diff
@@ -1,47 +1,52 @@
...
-    bell[1, 2] = bell[2, 1] = 1 / 2
-    bell[0, 0] = bell[3, 3] = 1 / 2
-    bell[1, 1] = bell[2, 2] = 1 / 2
-    bell = bell / np.trace(bell)
-    assert is_abs_ppt(bell, [2, 2]) is False
-
+@pytest.mark.parametrize(
+    "matrix, dims, expected_result",
+    [
+        # 2x2: Bell state (not absolutely PPT)
```
Comment (vprusso) [inline toqito/state_props/tests/test_is_abs_ppt.py:22]: [nit]: Please end comments with "`.`"

File: toqito/state_props/tests/test_is_abs_ppt.py [hunk 13]
```diff
@@ -1,47 +1,52 @@
...
+        # 3x3: Block-diagonal state with two large and one small block (not absolutely PPT)
+        (np.diag([0.45, 0.45, 0.05, 0.01, 0.01, 0.01, 0.01, 0.005, 0.005]), [3, 3], False),
+        # 3x3: State with one eigenvalue close to 1 (not absolutely PPT)
+        (np.diag([0.97, 0.01, 0.01, 0.003, 0.002, 0.002, 0.001, 0.001, 0.001]), [3, 3], False),
+    ],
+)
+def test_known_absppt_and_non_absppt(matrix, dims, expected_result):
+    """Test known absolutely PPT and non-absolutely PPT states for 2x2, 2x3, 3x3, and edge cases."""
+    # Normalize if not already
+    matrix = matrix / np.trace(matrix)
```
Comment (vprusso) [inline toqito/state_props/tests/test_is_abs_ppt.py:47]: Is normalization necessary for test cases that we define ourselves? Additionally, would this make sense in the actual function instead of in a test? It seems odd to normalize in the test only but not in the function.

File: toqito/state_props/tests/test_is_abs_ppt.py (hunk 11 above)
Comment (purva-thakre) [reply toqito/state_props/tests/test_is_abs_ppt.py:3]: @vprusso It is being used in a unit test. 

image

File: toqito/state_props/tests/test_is_abs_ppt.py (hunk 13 above)
Comment (luffy-orf) [reply toqito/state_props/tests/test_is_abs_ppt.py:47]: Good point! We now only normalize in tests if the matrix isn’t already trace 1. For hand-crafted test cases, we ensure the input is a valid density matrix, matching the function’s contract. The function itself does not normalize input, but validates it and raises an error if not trace 1.

File: toqito/state_props/is_abs_ppt.py [hunk 14]
```diff
@@ -2,19 +2,15 @@
...
 
-    Only implemented for 2x2, 2x3, and 3x3 systems.
+    :cite:`Hildebrand_2007_PPT, QETLAB_link`.
 
-    References
-    ==========
-    .. bibliography::
-        :filter: docname in docnames
-        :key: Hildebrand_2007_PPT, QETLAB_IsAbsPPT
+    Only implemented for 2x2, 2x3, and 3x3 systems.
```
[the remark this answers is not in the document]
Comment (luffy-orf) [reply toqito/state_props/is_abs_ppt.py:13]: The limitation to 2x2, 2x3, and 3x3 bipartite systems is due to the current state of the mathematical theory and available criteria for the absolutely PPT property

**Comment by luffy-orf** [comment]
@purva-thakre @vprusso , Please review the PR

File: toqito/state_props/is_abs_ppt.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+            [
+                [2 * eigvals[8], eigvals[7] - eigvals[0], eigvals[5] - eigvals[1]],
+                [eigvals[7] - eigvals[0], 2 * eigvals[6], eigvals[4] - eigvals[2]],
+                [eigvals[5] - eigvals[1], eigvals[4] - eigvals[2], 2 * eigvals[3]],
+            ]
+        )
+        constraints.extend([L1, L2])
+    else:
+        # Not implemented for higher dimensions
+        raise NotImplementedError("Absolutely PPT constraints for dimensions > 3x3 are not implemented.")
```
Comment (vprusso) [inline toqito/state_props/is_abs_ppt.py:70]: I believe we will want to implement the SDP that is also implemented in QETLAB here in this case, correct?
https://github.com/nathanieljohnston/QETLAB/blob/f589c35a29e03624ecfd972fd209014799d24862/IsAbsPPT.m#L95

**Comment by purva-thakre** [comment]
@luffy-orf Are you planning on completing this?
